### ISSUE TITLE
RISC-V Support?

### ISSUE THREAD

**Issue by timsnyder** [conda-forge/conda-forge.github.io#1744]
### Your question:

I'd like to start a discussion about adding limited support for RISC-V architectures.  I'm in the process of helping some colleagues at UC Berkeley transition to using conda-forge instead of system packages for https://fires.im/ and https://chipyard.readthedocs.io/en/stable/.  We've transitioned the system yum and pip packages to using conda-forge.  The next steps are:
1. to create packages for other x86 things we currently build from source (e.g. [libdwarf](https://github.com/conda-forge/staged-recipes/pull/18638) and [QEMU](https://github.com/conda-forge/staged-recipes/pull/18402))
2. migrate pre-compiled toolchain install https://chipyard.readthedocs.io/en/stable/Chipyard-Basics/Initial-Repo-Setup.html#building-a-toolchain (`--ec2-fast`) to using conda instead of the current tarball extraction technique that is limited in it's portability and has no infrastructure to help in it's maintenance.

Adding the toolchain makes me wonder -- does it make sense to start adding RISC-V architectures to conda-forge?   For starters, centos does not support it yet (Fedora does).  The current use model builds and installs https://github.com/riscv-collab/riscv-gnu-toolchain into a `$RISCV` prefix (many scripts use the `RISCV` environment variable) and then installs more RISC-V ABI software into that same prefix.  I assume that conda-forge would want the cross-toolchain to be installed into a `linux-64` "build prefix" and the subsequent software installed into a prefix that is architecture-specific (for the sake of simplicity, I'm only considering `linux-rv64gc` for now because there are many pieces and options when it comes to RISC-V isa).  Come to think of it, we may just want to call it `linux-rv64` and not have more details of which ISA pieces are actually supported encoded in the platform name.

Regardless of whether @conda-forge/core is interested in adding RISC-V platforms, to be able to upload to a new platform type on anaconda.org, we would need to get support for it from Anaconda, right?  How does that work when conda-forge supports a platform that upstream Anaconda doesn't? (e.g. `linux-aarch64`)

**Comment by chrisburr** [comment]
These two issues will probably help you get an idea of what's needed (especially https://github.com/conda-forge/conda-forge.github.io/issues/1123#issuecomment-668399522):

* https://github.com/conda-forge/conda-forge.github.io/issues/1123
* https://github.com/conda-forge/conda-forge.github.io/issues/1126

**Comment by chenghlee** [comment]
> ...able to upload to a new platform type on anaconda.org...

Just checked; anaconda.org already supports uploading of `linux-riscv64` conda packages.

**Comment by h-vetinari** [comment]
Following the initial [outreach](https://conda-forge.zulipchat.com/#narrow/channel/457337-general/topic/Bringing.20conda-forge.20to.20RISC-V.20.28We.20have.20a.20working.20miniforge/with/555198090) on zulip, the OpenKylin community presented their case for building up riscv64 in conda-forge in the most recent core call ([minutes](https://github.com/conda-forge/conda-forge.github.io/blob/main/community/minutes/2025-11-12.md)). While there are several open questions, the way I interpreted the overall impression of the call is that @conda-forge/core is generally interested and willing to find ways together to make this work.

Some open topics:
* Native hardware support:
  * apparently substantial resources can be provided for native CI. One possibility that was mentioned was plugging this into self-hosted github action runners. Not sure if this is in scope of cirun or would be done separately.
* ABI varieties of `riscv64`:
  *  Apparently this is mostly `rv64gc` and the newer `RVA23`; the latter is a superset of the former, so these are not divergent ABIs per se, just regular evolution/extension. Quoting from that Zulip thread: "The best example is Ubuntu: 24.04 was `rv64gc`, but 25.10 will be `RVA23`-only." I trust the OpenKylin team to make a recommendation here, but since we're starting from scratch, I think we can go with the newer instruction set that everyone is apparently moving to anyway.
* [something else I forgot?]

Some of these questions will have longer lead times (e.g. CI integration), but we don't need to solve everything at once. On a shorter time horizon, the first step is to bootstrap the necessary compilers. Copying the checklist from [here](https://github.com/conda-forge/conda-forge.github.io/issues/1123#issuecomment-668399522):

* [x] Ask for volunteers for conda-forge/help-riscv64 team to support riscv64. (At least two people who'll respond to github issues quickly)
  * OpenKylin community mentioned being able to have 2-3 people on this _full time_.
  * Please provide github handles for them.
* [x] Ask Anaconda, Inc to support uploading riscv64 packages to anaconda.org.
  * as mentioned by Cheng above
* [ ] Send a PR to repackage the centos sysroot for riscv64 in https://github.com/conda-forge/linux-sysroot-feedstock/
  * I'm working on unblocking https://github.com/conda-forge/linux-sysroot-feedstock/pull/92 as a prerequisite
  * However, I've noticed that almalinux _**does not**_ (yet) support riscv[^1], i.e. we cannot use their RPMs for this.
  * As it turns out, rocky linux [added](https://rockylinux.org/news/rockylinux-support-for-riscv) riscv support for rocky v10; this should be pretty compatible, but will need more work due to onboarding another distro flavour
* [ ] Build binutils for riscv64 in https://github.com/conda-forge/binutils-feedstock
* [ ] Build a cross compiler for riscv64 in https://github.com/conda-forge/ctng-compilers-feedstock
* [ ] Build a cross-native compiler for riscv64 in https://github.com/conda-forge/ctng-compilers-feedstock

[^1]: check content of https://repo.almalinux.org/almalinux/10.0/BaseOS, for example. I found an old post on their [forums](https://forums.almalinux.org/t/add-support-for-riscv64/3402), but it has 0 responses.

**Comment by bitsk** [comment]
Hi,@h-vetinari! Here are our openKylin participants: Tianxiang Bian @bitsk , Jiankang Li @lijkang, Ruian Zhan @scorbunny-cinderace

**Comment by bitsk** [comment]
We need to update the QEMU version in [docker-images](https://github.com/conda-forge/docker-images/blob/main/download-qemu-static.sh) to 10 to support RVA23([SE_01_021 - QEMU RVA23 profile support](https://lf-rise.atlassian.net/wiki/spaces/HOME/pages/8586734/SE_01_021+-+QEMU+RVA23+profile+support)).

However, the QEMU 10 packages in the [Fedora repository (e.g., 10.1.2)](https://kojipkgs.fedoraproject.org/packages/qemu/10.1.2/)are built for fc43 and fc44 and may not run on alma or rocky.

**Comment by h-vetinari** [comment]
> However, the QEMU 10 packages in the [Fedora repository (e.g., 10.1.2)](https://kojipkgs.fedoraproject.org/packages/qemu/10.1.2/) are built for fc43 and fc44 and may not run on alma or rocky.

You might be interested in https://github.com/conda-forge/docker-images/issues/320 ;-)

**Comment by h-vetinari** [comment]
https://github.com/conda-forge/linux-sysroot-feedstock/pull/93 has been merged. @bitsk, @lijkang, @scorbunny-cinderace, are you planning to tackle binutils and the compilers? I'm not really planning to do that myself. Help on the qemu side is also welcome, though there at least it's on my list of planned things, because it'll be useful in any case.

**Comment by bitsk** [comment]
The GCC PR is available here: https://github.com/conda-forge/ctng-compilers-feedstock/pull/205

**Comment by bitsk** [comment]
Hi @h-vetinari @isuruf, just a quick update on our recent progress.

We expect to receive our first batch of **RVA23** development boards (8+8 cores) in about **two weeks**. We’ve just had a chance to test them, and the performance has exceeded our expectations. Additionally, we’ve made significant progress with our RISC-V porting on **OpenStack**, which is now running successfully on the UltraRISC RISC-V platform.

I would also like to introduce Chenguang Ding (@Divinity96) and Zhuoheng Li (@zohanzephyr) from KylinSoft, who will be joining our subsequent efforts. They will be focusing on patches for base software and fixing build issues.

**Comment by bitsk** [comment]
> Trying to access http://conda.riscv.uno/ but it's not working, is it just me?

@luhenry The channel has moved to https://conda.openkylin.top/index.html

**Comment by luhenry** [comment]
I also got a more recent build of micromamba (fully statically linked) available: https://github.com/riseproject-dev/micromamba-releases-mirror/releases/tag/2.5.0-1. Built with https://github.com/riseproject-dev/micromamba-releases-mirror/blob/main/.github/workflows/build-micromamba.yml on native hardware

**Comment by luhenry** [comment]
Just went through the exercise of listing all the packages we need to exist on [`conda.anaconda.org/conda-forge/`](https://conda.anaconda.org/conda-forge/) to get a working `linux-anvil-riscv64` docker image (as of April 8th, 2026).

I broke it down into 4 layers:
- **Layer 0:** what's needed to get cross-compilation going ([listed above](https://github.com/conda-forge/conda-forge.github.io/issues/1744#issuecomment-3537124401))
- **Layer 1:** what's needed to get the [micromamba-feedstock](https://github.com/conda-forge/micromamba-feedstock/blob/main/recipe/meta.yaml) built
- **Layer 2:** dependencies of Miniforg3 as listed in [miniforge/Miniforge3/construct.yaml](https://github.com/conda-forge/miniforge/blob/main/Miniforge3/construct.yaml)
- **Layer 3:** packages installed in the `linux-anvil-riscv64` docker image as listed in [docker-images/scripts/run_commands](https://github.com/conda-forge/docker-images/blob/a63fbbb9038abc120fd55ee47973936eb382590b/scripts/run_commands#L67-L75)

---

## Layer 0: toolchains

Built on `linux-64`, targeting `riscv64`. The binaries run on x86_64 or riscv64 but produce output for riscv64.

| | Package | Subdir | Feedstock | Package | PR |
|---|---|---|---|---|---|
| ✅ | `binutils_impl_linux-riscv64`          | `linux-64`      | [binutils-feedstock](https://github.com/conda-forge/binutils-feedstock)                                 | [anaconda.org](https://anaconda.org/conda-forge/binutils_impl_linux-riscv64)   | |
| ✅ | `binutils_impl_linux-riscv64`          | `linux-riscv64` | [binutils-feedstock](https://github.com/conda-forge/binutils-feedstock)                                 | [anaconda.org](https://anaconda.org/conda-forge/binutils_impl_linux-riscv64)   | ~[#114](https://github.com/conda-forge/binutils-feedstock/pull/114)~ [#116](https://github.com/conda-forge/binutils-feedstock/pull/116) |
| ✅ | `binutils_linux-riscv64`               | `linux-64`      | [binutils-feedstock](https://github.com/conda-forge/binutils-feedstock)                                 | [anaconda.org](https://anaconda.org/conda-forge/binutils_linux-riscv64)        | |
| ✅ | `binutils_linux-riscv64`               | `linux-riscv64` | [binutils-feedstock](https://github.com/conda-forge/binutils-feedstock)                                 | [anaconda.org](https://anaconda.org/conda-forge/binutils_linux-riscv64)        | ~[#114](https://github.com/conda-forge/binutils-feedstock/pull/114)~ [#116](https://github.com/conda-forge/binutils-feedstock/pull/116) |
| ✅ | `gcc_impl_linux-riscv64`               | `linux-64`      | [ctng-compilers-feedstock](https://github.com/conda-forge/ctng-compilers-feedstock)                     | [anaconda.org](https://anaconda.org/conda-forge/gcc_impl_linux-riscv64)        | [#205](https://github.com/conda-forge/ctng-compilers-feedstock/pull/205) |
| ✅ | `gcc_impl_linux-riscv64`               | `linux-riscv64` | [ctng-compilers-feedstock](https://github.com/conda-forge/ctng-compilers-feedstock)                     | [anaconda.org](https://anaconda.org/conda-forge/gcc_impl_linux-riscv64)        | [#216](https://github.com/conda-forge/ctng-compilers-feedstock/pull/216) |
| ✅ | `gcc_linux-riscv64`                    | `linux-64`      | [ctng-compiler-activation-feedstock](https://github.com/conda-forge/ctng-compiler-activation-feedstock) | [anaconda.org](https://anaconda.org/conda-forge/gcc_linux-riscv64)             | [#189](https://github.com/conda-forge/ctng-compiler-activation-feedstock/pull/189), [#192](https://github.com/conda-forge/ctng-compiler-activation-feedstock/pull/192), [admin-requests#2125](https://github.com/conda-forge/admin-requests/pull/2125) |
| ✅ | `gcc_linux-riscv64`                    | `linux-riscv64` | [ctng-compiler-activation-feedstock](https://github.com/conda-forge/ctng-compiler-activation-feedstock) | [anaconda.org](https://anaconda.org/conda-forge/gcc_linux-riscv64)             | [#197](https://github.com/conda-forge/ctng-compiler-activation-feedstock/pull/197) |
| ✅ | `gendef`                               | `linux-riscv64` | [gendef-feedstock](https://github.com/conda-forge/gendef-feedstock)                                     | [anaconda.org](https://anaconda.org/conda-forge/gendef)                        | [#8](https://github.com/conda-forge/gendef-feedstock/pull/8) |
| ✅ | `gfortran_impl_linux-riscv64`          | `linux-64`      | [ctng-compilers-feedstock](https://github.com/conda-forge/ctng-compilers-feedstock)                     | [anaconda.org](https://anaconda.org/conda-forge/gfortran_impl_linux-riscv64)   | [#205](https://github.com/conda-forge/ctng-compilers-feedstock/pull/205) |
| ✅ | `gfortran_impl_linux-riscv64`          | `linux-riscv64` | [ctng-compilers-feedstock](https://github.com/conda-forge/ctng-compilers-feedstock)                     | [anaconda.org](https://anaconda.org/conda-forge/gfortran_impl_linux-riscv64)   | [#216](https://github.com/conda-forge/ctng-compilers-feedstock/pull/216) |
| ✅ | `gfortran_linux-riscv64`               | `linux-64`      | [ctng-compiler-activation-feedstock](https://github.com/conda-forge/ctng-compiler-activation-feedstock) | [anaconda.org](https://anaconda.org/conda-forge/gfortran_linux-riscv64)        | [#189](https://github.com/conda-forge/ctng-compiler-activation-feedstock/pull/189), [#192](https://github.com/conda-forge/ctng-compiler-activation-feedstock/pull/192), [admin-requests#2125](https://github.com/conda-forge/admin-requests/pull/2125) |
| ✅ | `gfortran_linux-riscv64`               | `linux-riscv64` | [ctng-compiler-activation-feedstock](https://github.com/conda-forge/ctng-compiler-activation-feedstock) | [anaconda.org](https://anaconda.org/conda-forge/gfortran_linux-riscv64)        | [#197](https://github.com/conda-forge/ctng-compiler-activation-feedstock/pull/197) |
| ✅ | `gxx_impl_linux-riscv64`               | `linux-64`      | [ctng-compilers-feedstock](https://github.com/conda-forge/ctng-compilers-feedstock)                     | [anaconda.org](https://anaconda.org/conda-forge/gxx_impl_linux-riscv64)        | [#205](https://github.com/conda-forge/ctng-compilers-feedstock/pull/205) |
| ✅ | `gxx_impl_linux-riscv64`               | `linux-riscv64` | [ctng-compilers-feedstock](https://github.com/conda-forge/ctng-compilers-feedstock)                     | [anaconda.org](https://anaconda.org/conda-forge/gxx_impl_linux-riscv64)        | [#216](https://github.com/conda-forge/ctng-compilers-feedstock/pull/216) |
| ✅ | `gxx_linux-riscv64`                    | `linux-64`      | [ctng-compiler-activation-feedstock](https://github.com/conda-forge/ctng-compiler-activation-feedstock) | [anaconda.org](https://anaconda.org/conda-forge/gxx_linux-riscv64)             | [#189](https://github.com/conda-forge/ctng-compiler-activation-feedstock/pull/189), [#192](https://github.com/conda-forge/ctng-compiler-activation-feedstock/pull/192), [admin-requests#2125](https://github.com/conda-forge/admin-requests/pull/2125) |
| ✅ | `gxx_linux-riscv64`                    | `linux-riscv64` | [ctng-compiler-activation-feedstock](https://github.com/conda-forge/ctng-compiler-activation-feedstock) | [anaconda.org](https://anaconda.org/conda-forge/gxx_linux-riscv64)             | [#197](https://github.com/conda-forge/ctng-compiler-activation-feedstock/pull/197) |
| ✅ | `kernel-headers_linux-riscv64`         | `noarch`        | [linux-sysroot-feedstock](https://github.com/conda-forge/linux-sysroot-feedstock)                       | [anaconda.org](https://anaconda.org/conda-forge/kernel-headers_linux-riscv64)  | |
| ✅ | `ld_impl_linux-riscv64`                | `linux-64`      | [binutils-feedstock](https://github.com/conda-forge/binutils-feedstock)                                 | [anaconda.org](https://anaconda.org/conda-forge/ld_impl_linux-riscv64)         | |
| ✅ | `ld_impl_linux-riscv64`                | `linux-riscv64` | [binutils-feedstock](https://github.com/conda-forge/binutils-feedstock)                                 | [anaconda.org](https://anaconda.org/conda-forge/ld_impl_linux-riscv64)         | ~[#114](https://github.com/conda-forge/binutils-feedstock/pull/114)~ [#116](https://github.com/conda-forge/binutils-feedstock/pull/116) |
| ✅ | `libgcc-devel_linux-riscv64`           | `noarch`        | [ctng-compilers-feedstock](https://github.com/conda-forge/ctng-compilers-feedstock)                     | [anaconda.org](https://anaconda.org/conda-forge/libgcc-devel_linux-riscv64)    | [#205](https://github.com/conda-forge/ctng-compilers-feedstock/pull/205) |
| ✅ | `libstdcxx-devel_linux-riscv64`        | `noarch`        | [ctng-compilers-feedstock](https://github.com/conda-forge/ctng-compilers-feedstock)                     | [anaconda.org](https://anaconda.org/conda-forge/libstdcxx-devel_linux-riscv64) | [#205](https://github.com/conda-forge/ctng-compilers-feedstock/pull/205) |
| 🚧 | `rust_linux-riscv64`                   | `linux-64`      | [rust-activation-feedstock](https://github.com/conda-forge/rust-activation-feedstock)                                         | [anaconda.org](https://anaconda.org/conda-forge/rust_linux-riscv64)            | [#104](https://github.com/conda-forge/rust-activation-feedstock/pull/104), [admin-requests#2259](https://github.com/conda-forge/admin-requests/pull/2259) |
| 🚧 | `rust_linux-riscv64`                   | `linux-riscv64` | [rust-activation-feedstock](https://github.com/conda-forge/rust-activation-feedstock)                                         | [anaconda.org](https://anaconda.org/conda-forge/rust_linux-riscv64)            | [#104](https://github.com/conda-forge/rust-activation-feedstock/pull/104), [admin-requests#2259](https://github.com/conda-forge/admin-requests/pull/2259) |
| ✅ | `rust-std-riscv64gc-unknown-linux-gnu` | `noarch`        | [rust-feedstock](https://github.com/conda-forge/rust-feedstock)                                         | [anaconda.org](https://anaconda.org/conda-forge/rust-std-riscv64gc-unknown-linux-gnu) | [#297](https://github.com/conda-forge/rust-feedstock/pull/297), [admin-requests#2244](https://github.com/conda-forge/admin-requests/pull/2244) |
| ✅ | `sysroot_linux-riscv64`                | `noarch`        | [linux-sysroot-feedstock](https://github.com/conda-forge/linux-sysroot-feedstock)                       | [anaconda.org](https://anaconda.org/conda-forge/sysroot_linux-riscv64)         | |

### Pinning
* https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/8721
* https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/8727

### Migration
* https://github.com/conda-forge/conda-forge-bot/pull/6316
* https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/8787
* https://conda-forge.org/status/migration/?name=supportlinuxriscv64platform

@bitsk you're working on the Layer 0 at the moment. Does the rest makes sense to you?

**Comment by bitsk** [comment]
> @bitsk you're working on the Layer 0 at the moment. Does the rest makes sense to you?

@luhenry  Yes, we are currently waiting for the new qemu solution (https://github.com/conda-forge/qemu-execve-feedstock/pull/15) to be integrated into the docker build images (https://github.com/conda-forge/docker-images/issues/320). Since our initial goal was to push for RVA23, a version 10 of qemu was required. However, following our previous [discussion](https://hackmd.io/@conda-community/SJ_oFitUZg), we decided to postpone RVA23, so is your current PR (https://github.com/conda-forge/docker-images/pull/328) acceptable now?

@h-vetinari , do we have any recent progress on qemu? Once the base toolchain issues are resolved, we will quickly move on to the next phase of the build.

**Comment by luhenry** [comment]
> However, following our previous [discussion](https://hackmd.io/@conda-community/SJ_oFitUZg), we decided to postpone RVA23, so is your current PR ([conda-forge/docker-images#328](https://github.com/conda-forge/docker-images/pull/328)) acceptable now?

IIUC, for this PR to be merged, we first need to have builds of micromamba and miniforge3, but to get these builds, we need all their dependencies to be available in the `linux-riscv64` channel? Or are we able to cross-compile these two and so we are only left with the packages that are installed in docker-images’s [run_commands](https://github.com/conda-forge/docker-images/blob/a63fbbb9038abc120fd55ee47973936eb382590b/scripts/run_commands#L67-L75)?

Question, not trying to rehash the debate that happened previously, if we target rv64gc now, how hard is it to add/switch to RVA23 later? Is there support for different binaries like x86 v1/v2/v3/v4? Or is it “we target the most baseline for maximum compatibility”?

> [ h-vetinari](https://github.com/h-vetinari) , do we have any recent progress on qemu? Once the base toolchain issues are resolved, we will quickly move on to the next phase of the build.

I am working on adding support for Azure DevOps in the [RISE RISC-V Runners](https://riseproject-dev.github.io/riscv-runner/), so that, if you want, you can have easy access to a large pool of riscv64 runners. (Ignore the exclusive mention of GitHub, it will be extended to other CI/CD systems)

**Comment by isuruf** [comment]
> we need all their dependencies to be available in the linux-riscv64 channel?

No

> Or are we able to cross-compile these two and so we are only left with the packages that are installed in docker-images’s [run_commands](https://github.com/conda-forge/docker-images/blob/a63fbbb9038abc120fd55ee47973936eb382590b/scripts/run_commands#L67-L75)?

Yes, everything there can be cross compiled.

**Comment by bitsk** [comment]
>how hard is it to add/switch to RVA23 later?

I don't think the transition to RVA23 in conda-forge will be very difficult. Based on our experience building distributions, most issues stem from stricter compilation checks introduced by GCC version upgrades. Since we are directly using GCC 15, there shouldn't be too many of these issues. Regarding dynamic feature detection, there probably isn't such a mechanism available right now, but I believe the RVA23 baseline is sufficient.

**Comment by bitsk** [comment]
cc @isuruf @h-vetinari @conda-forge/help-riscv64

Quick update: conda-forge/ctng-compilers-feedstock#205 has been merged, so the GCC cross-compilation toolchain for `linux-riscv64` is now ready.

The native GCC toolchain still needs some `linux-riscv64` dependencies to be built first. Do you think the next step should be to start an automated `linux-riscv64` migration so these dependencies can be built in batch, or would you recommend bootstrapping them manually first?

**Comment by h-vetinari** [comment]
We'll still need a PR to https://github.com/conda-forge/ctng-compiler-activation-feedstock, but that should be much easier, because it's only a meta-package.

**Comment by luhenry** [comment]
The change to [conda-forge/ctng-compiler-activation-feedstock](https://github.com/conda-forge/ctng-compiler-activation-feedstock) was [merged](https://github.com/conda-forge/ctng-compiler-activation-feedstock/pull/189). Thanks @bitsk !

**Comment by isuruf** [comment]
> would you recommend bootstrapping them manually first?

We should do the bootstrapping manually until we get the native compilers built.

**Comment by luhenry** [comment]
@bitsk @isuruf what's the latest on this issue? I couldn't find any more PRs around that topic. Any way I could be helping (resources, funding, machines)?

**Comment by isuruf** [comment]
We need a PR to https://github.com/conda-forge/ctng-compilers-feedstock to build the native compilers.

**Comment by luhenry** [comment]
> We need a PR to https://github.com/conda-forge/ctng-compilers-feedstock to build the native compilers.

Looking at it now. I'll submit and link a draft PR.

Would you also be open to use the [RISE RISC-V Runners](https://riscv-runners.riseproject.dev/) for native runners on GitHub Actions? It's by [RISE](https://riseproject.dev/), a Linux Foundation project, focused on enabling the RISC-V Software Ecosystem. I'm the one maintaining it so happy to answer any questions.

**Comment by isuruf** [comment]
Yes, but we need to advance quite a bit before we are able to use the native runners.

**Comment by luhenry** [comment]
@isuruf IIUC now that you've merged https://github.com/conda-forge/ctng-compiler-activation-feedstock/pull/192 and [gcc_linux-riscv64](https://anaconda.org/channels/conda-forge/packages/gcc_linux-riscv64/overview), [gxx_linux-riscv64](https://anaconda.org/channels/conda-forge/packages/gxx_linux-riscv64/overview), and [gfortran_linux-riscv64](https://anaconda.org/channels/conda-forge/packages/gfortran_linux-riscv64/overview) are available on linux-x64, we should now build the "native" packages for `binutils_linux-riscv64`, `binutils_impl_linux-riscv64`, and `ld_impl_linux-riscv64`? I assume that's in https://github.com/conda-forge/binutils-feedstock?

@bitsk submitted https://github.com/conda-forge/binutils-feedstock/pull/114 already

**Comment by isuruf** [comment]
@luhenry, please do not make PRs on your own to add linux-riscv64 support for feedstocks. We need to make a migrator and let the bot make these PRs

**Comment by luhenry** [comment]
One thing I noticed from the few PRs I (wrongfully) opened is that it fails on a similar error [1] [2] [3]:
```
+++ micromamba create -p /opt/conda/envs/sysroot_linux-riscv64 --root-prefix /home/conda/.conda --yes --quiet --override-channels --channel conda-forge --strict-channel-priority sysroot_linux-riscv64=2.34
error    libmamba Could not solve for environment specs
    The following package could not be installed
    └─ sysroot_linux-riscv64 2.34**  does not exist (perhaps a typo or a missing channel).
```

As far as I understand `recipe/conda_build_config.yaml` in [conda-forge/conda-forge-pinning-feedstock](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/main/recipe/conda_build_config.yaml#L21), it should be 2.39 and not 2.34? What am I missing here?

[1]: https://github.com/luhenry/conda-forge-reproc-feedstock/actions/runs/29377747632/job/87234625466#step:4:375
[2]: https://github.com/luhenry/conda-forge-xz-feedstock/actions/runs/29378144899/job/87235759203#step:4:380
[3]: https://github.com/luhenry/conda-forge-curl-feedstock/actions/runs/29377947047/job/87235208048#step:4:376

**Comment by luhenry** [comment]
@isuruf @h-vetinari, now that we have `gcc_linux-riscv64`/`gxx_linux-riscv64` on `linux-riscv64` [1], what are the next steps? Is it when the migrator [2] comes into play, or are there any other projects we would need to do before?

[1]: https://anaconda.org/channels/conda-forge/packages/gcc_linux-riscv64/overview
[2]: https://github.com/conda-forge/conda-forge-bot/pull/6316

**Comment by isuruf** [comment]
Can you send a manual PR to zlib-feedstock? That should tell us if we need anything more?

**Comment by luhenry** [comment]
It works [1] ! And the change: https://github.com/conda-forge/zlib-feedstock/pull/93

[1]: https://github.com/conda-forge/zlib-feedstock/actions/runs/30397997671/job/90406786370?pr=93

**Comment by h-vetinari** [comment]
@luhenry, I put a link to your comment at the top for ease of reference; could you please keep individual tables (per section) ordered alphabetically?

**Comment by h-vetinari** [comment]
Note that a big milestone has been achieved here: the [risc-v migration](https://conda-forge.org/status/migration/?name=supportlinuxriscv64platform) has started with an initial set of packages. Thanks to everyone who helped push for this, and apologies it took longer than many hoped initially.

CC @conda-forge/help-riscv64

**Comment by isuruf** [comment]
@conda-forge/help-riscv64, please have a look at all the open PRs and help out if the CI fails for linux-riscv64 platform.

**Comment by h-vetinari** [comment]
Biggest issue right now is a cycle between `pkg-config` - `glib` - `python`; the `pkg-config` can nominally build without `glib`, but in that case, the build falls back to a vendored internal copy, which doesn't survive _its_ `config` call in cross-compilation; `glib` in turn requires python, and `python` requires `pkg-config`.
* https://github.com/conda-forge/pkg-config-feedstock/pull/47
* https://github.com/conda-forge/glib-feedstock/pull/245

`pkg-config` also blocks a bunch of packages with many children (e.g. `gtest` & `icu`), although the migrator does not capture this in the dependency graph.

**Comment by luhenry** [comment]
@h-vetinari the solution is then to fix `pkg-config` and make sure its vendored internal copy do survive its `config`?

**Comment by h-vetinari** [comment]
That sounds like the easiest approach, hopefully. Feel free to give it a shot! :)

**Comment by isuruf** [comment]
Building a `python[subdir=linux-risc64]` does not require a `pkg-config[subdir=linux-risc64]`. It only requires a `pkg-config[linux-64]`. Therefore there's no cycle.

**Comment by h-vetinari** [comment]
Seeing https://github.com/conda-forge/libcurand-feedstock/pull/33, I was wondering where CUDA stands w.r.t. risc-v, I found some announcements that this is planned, but I don't think there's anything concrete yet? CC @conda-forge/cuda

Also worth noting: CUDA on risc-v will [require](https://riscv.org/blog/nvidia-cuda-rva23/) the rva23 profile.

**Comment by h-vetinari** [comment]
> Building a `python[subdir=linux-risc64]` does not require a `pkg-config[subdir=linux-risc64]`. It only requires a `pkg-config[linux-64]`. Therefore there's no cycle.

That's true, I had only done a quick grep. Of the three, python has the widest set of dependencies though, so it might still get stuck on something else on the way there. Overall, if we can get `pkg-config` to work with its vendored glib, that's probably still the easiest way to move forward

**Comment by carterbox** [comment]
@h-vetinari, your assessment is correct. NVIDIA has announced that we are working to bring CUDA to RISC-V host platforms; however it is not yet available in any toolkit.

**Comment by bitsk** [comment]
> [@conda-forge/help-riscv64](https://github.com/orgs/conda-forge/teams/help-riscv64), please have a look at all the open PRs and help out if the CI fails for linux-riscv64 platform.

On it! We've started reviewing the open PRs and monitoring the linux-riscv64 CI status.

**Comment by bitsk** [comment]
Regarding `pkg-config`, in our preview builds we temporarily dropped the glib dependency to get it to build successfully. We could put out a version like this first, and then re-add the glib dependency later.

**Comment by bitsk** [comment]
We may have found a migration bookkeeping issue:

[`zlib`](https://github.com/conda-forge/zlib-feedstock/pull/93) and [`gendef`](https://anaconda.org/channels/conda-forge/packages/gendef/files) have no unfinished parents and already provide successful `linux-riscv64` packages, but the [migration page](https://conda-forge.org/status/migration/?name=supportlinuxriscv64platform) still shows them as `awaiting-parents`.

They were completed manually before this migration started, so they may be missing the corresponding `LinuxRISCV64` migration records, leaving downstream packages such as `zstd` blocked.

@isuruf @h-vetinari , could these two nodes be marked as completed in the bot backend? This would avoid duplicate PRs and unblock the next migration layer.

**Comment by h-vetinari** [comment]
I noticed the same, see my post-merge comment in https://github.com/conda-forge/conda-forge-bot/pull/6419

**Comment by luhenry** [comment]
@isuruf @h-vetinari @wolfv `xorg-libx11` is [marked as done](https://conda-forge.org/status/migration/?name=supportlinuxriscv64platform), but there was [no pull requests](https://github.com/conda-forge/xorg-libx11-feedstock/pulls?q=is%3Apr) to enable it, it's [not in the `conda-forge.yml` file](https://github.com/conda-forge/xorg-libx11-feedstock/blob/c396d0088712df58159ec14017da16964b9371db/conda-forge.yml), and so it's [not in a package either](https://anaconda.org/channels/conda-forge/packages/xorg-libx11/overview). How do we reset the state of the migrator for that specific package? Will it eventually converge in some ways?

**Comment by pkgw** [comment]
I've just been looking at the same packages — `libxcb` and `xcb-proto` are both in the same state as x11: marked as done but no PR was ever filed. FWIW, in the raw migration status file I see, e.g.:

```json
"pr_url": "https://github.com/conda-forge/xcb-proto-feedstock"
```

I took a quick look at doing a manual migration and `xcb-proto` can't go because it wants `cross-python_linux-riscv64`, which is also what `glib` is now stuck on. I don't see `cross-python` in the migration plan at all — does that also need to be done manually?

**Comment by h-vetinari** [comment]
There were several feedstocks marked as done even though no PR had been merged; AFAIU from discussion in https://github.com/conda-forge/conda-forge-bot/pull/6439, this is likely due to a "cycle breaker"; of which there were a few (involving llvmdev, ncurses, python and various xorg libs).

Telling the bot to redo nodes it thinks it has done already is tricky, short of restarting the migration. For now I suggest to try opening manual PRs for feedstocks that wrongly got marked done.

**Comment by luhenry** [comment]
> Telling the bot to redo nodes it thinks it has done already is tricky, short of restarting the migration. For now I suggest to try opening manual PRs for feedstocks that wrongly got marked done.

I'll have a look, going recursive wherever it needs, and linking back to your comment @h-vetinari for context.

**Comment by h-vetinari** [comment]
I opened a dedicated issue for this on the bot repo, which will make it easier to track this, and will not unnecessarily overload this issue with xref's.

**Comment by pkgw** [comment]
Thanks for figuring what's going on there.

The packages that I'm most involved with are now blocked on `cross-python`, which I don't see in the migration plan. Is there expected to be anything unusual about that package for riscv64?

**Comment by isuruf** [comment]
Tests are not being run in RISC-V platform (and all other linux cross compiled platforms) because of a bug in rattler-build. See https://github.com/prefix-dev/rattler-build/issues/2587#issuecomment-5296577480

v0 recipes with conda-build do run the tests.

**Comment by h-vetinari** [comment]
> Following projects were marked done when they weren't:

Let's please use https://github.com/conda-forge/conda-forge-bot/issues/6479 for this, to avoid overloading this thread

**Comment by isuruf** [comment]
Until https://github.com/prefix-dev/rattler-build/issues/2587#issuecomment-5296577480 is fixed, shall we hold off on the risc-v migration? It's really bad that we are not testing the packages due to this bug.

**Comment by h-vetinari** [comment]
I'd prefer to continue, for a couple of main reasons:
* the effort to do v1 conversions is unrelated to the riscv migration. @wolfv took the opportunity to convert a couple of feedstocks on the occasion of the migration, but it's not like one needs the other.
* most cross-compiled recipes are fine with skipping tests in build scripts[^1], so the existence of `CROSSCOMPILE_EMULATOR` etc. is secondary.
* the problem about variable detection is a philosophical discussion where both sides have strong conviction in diametrically opposite positions (rattler-build that doesn't want to parse build scripts, vs. keeping the same behaviour as conda-build). Resolving such entrenched differences can be a long and messy process, and we shouldn't let the people interested in riscv support suffer from gridlock on a completely unrelated topic.

[^1]: e.g by using `if [[ $build_platform == $target_platform ]]; then` guards around the test invocation

I'd prefer that we get to a working set of native build tools ASAP, and then figure out how to switch to native runners.

Finally, for recipes that you care about, we can undo the v1 conversion for now, and then those packages will again be tested as usual.

**Comment by h-vetinari** [comment]
With https://github.com/conda-forge/python-feedstock/pull/896 nearing merge, it looks like we have broken through the biggest set of missing dependencies, and things might go relatively quickly (🤞) until we get the native toolchains built out.

My question to @conda-forge/help-riscv64 is: what's the plan for native agents? There was an offer from the OpenKylin folks to provide runners, but I'm not sure if this is still current? Whatever the case may be, GHA doesn't offer riscv runners, so I assume we'll have to do _some_ integration work for that. It might be a good time to start thinking about developing and testing that setup already before we have all the packages in place.

**Comment by bitsk** [comment]
We can provide an OpenStack-based build cluster, and we are currently enabling public internet access for it. Do you have any recommendations regarding the access method?

**Comment by luhenry** [comment]
[RISE](https://riseproject.dev/) (Linux Foundation project) is also providing free to use [RISC-V Runners](https://riscv-runners.riseproject.dev/). It's a [GitHub App install](https://github.com/apps/rise-risc-v-runners) away (with limited credentials: self-hosted runners (read/write) only). It's being used by llama.cpp, Pytorch, CNCF, and some other projects. We are adding some RVA23-enabled hardware as well which will should be 2-3x faster than the current generation.

**Comment by velonica0** [comment]
Hi, @h-vetinari @conda-forge/help-riscv64

We(openKylin) currently offer a cluster based on the UR DP-1000, whose single-core performance can exceed that of the Spacemit K3. 
We will subsequently introduce a large number of K3s to provide support for RV23. 
Our hardware also integrates ROCm, validating the inference tasks of AMD graphics cards on the UR DP-1000.

**Comment by pavelzw** [comment]
Older versions of pixi/rattler-based tools take 2.28 as default glibc:

```
❯ pixi exec --spec rattler=0.2.11 -- rattler solve rattler-build --platform linux-riscv64
Solving for platform: linux-riscv64
INFO rattler_virtual_packages: cannot detect the version of the virtual package '__linux' when targeting 'linux-riscv64', assuming version 4.18; set the CONDA_OVERRIDE_LINUX environment variable to override
INFO rattler_virtual_packages: cannot detect the version of the virtual package '__glibc' when targeting 'linux-riscv64', assuming version 2.28; set the CONDA_OVERRIDE_GLIBC environment variable to override
Virtual packages:
  - __unix=0=0
  - __linux=4.18=0
  - __glibc=2.28=0
  - __archspec=1=riscv64

Error:   × Cannot solve the request because of: rattler-build * cannot be installed because there are no viable options:
  │ └─ rattler-build 0.74.0 | 0.75.0 would require
  │    └─ __glibc >=2.39,<3.0.a0, for which no candidates were found.
  │
```

https://github.com/conda/rattler/pull/2731, https://github.com/prefix-dev/pixi/pull/6914, https://github.com/prefix-dev/rattler-build/pull/2775  changes the default to 2.39. also added linux-riscv64 support in https://github.com/prefix-dev/setup-pixi/releases/tag/v0.10.2

**Comment by h-vetinari** [comment]
Heads-up: all the python-version dependent feedstocks that already have riscv PRs already (which is a lot, given that the bot marked python as done long ago; c.f. the xrefs in https://github.com/conda-forge/python-feedstock/pull/896) will have to be rerendered after https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/8897

**Comment by luhenry** [comment]
It works!
```
(base) root@256803bf8131:/# conda install python
Retrieving notices: done
Channels:
 - conda-forge
Platform: linux-riscv64
Collecting package metadata (repodata.json): done
Solving environment: done

# All requested packages already installed.
```

That's after installing Miniforge from https://github.com/conda-forge/miniforge/releases/tag/26.7.2-0. Congrats everyone! :tada:

**Comment by luhenry** [comment]
As we are getting closer to have native builds on RISC-V, I took a quick look at what packages are the most downloaded from conda-forge. Using [`condastats`](https://github.com/conda-incubator/condastats), I got that information for the month of `2026-08`: https://gist.github.com/luhenry/341794d80a9e325fc3a96d243f7e25e3.

Once we have the native runners, we can go through that list by adding them in batch to [`linux_riscv64.txt`](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/main/recipe/migration_support/linux_riscv64.txt)

**Comment by luhenry** [comment]
We now have the [linux-anvil-riscv64](https://hub.docker.com/r/condaforge/linux-anvil-riscv64) docker image available :tada:

**Comment by h-vetinari** [comment]
How large is the fleet of runners from RISE? @luhenry 

We haven't even [switched](https://github.com/conda-forge/conda-forge-bot/pull/6831) to native builds yet, but already I'm seeing long waits (>10min and climbing) for runners to be provisioned - there can't be more than a handful of feedstocks being tested on these machines yet (I'm aware of 3 that have unmerged PRs, and only [one](https://github.com/conda-forge/numpy-feedstock/pull/401) of those has CI running right now), so this would be a pretty major bottleneck.

**Comment by luhenry** [comment]
> How large is the fleet of runners from RISE? [@luhenry](https://github.com/luhenry)
> 
> We haven't even [switched](https://github.com/conda-forge/conda-forge-bot/pull/6831) to native builds yet, but already I'm seeing long waits (>10min and climbing) for runners to be provisioned - there can't be more than a handful of feedstocks being tested on these machines yet (I'm aware of 3 that have unmerged PRs, and only [one](https://github.com/conda-forge/numpy-feedstock/pull/401) of those has CI running right now), so this would be a pretty major bottleneck.

There are 55 machines right now. I've pushed quite a few changes in https://github.com/riseproject-dev which has led to higher peak than usual. I'll reduce the max usage of riseproject-dev so that there is more breathing room for other orgs.

Increased limit for conda-forge: https://github.com/riseproject-dev/riscv-runner/commit/7394c8b27b59396875d2d28f53c0f2a42754c338

**Comment by h-vetinari** [comment]
FWIW, https://github.com/conda-forge/coreutils-feedstock/pull/37 was built fully natively, so that's a good sign. However, I want to flag the following difference in runtimes across architectures (ppc is cross-compiled, so not apples-to-apples, but irrelevant for the overall job runtime)

<img width="440" height="420" alt="Image" src="https://github.com/user-attachments/assets/edf38a9b-701c-4923-860c-a17a5707eb30" />


That kind of differential between runners architectures will make it quite painful to progress on broader migrations (e.g. for new python versions), to say nothing about the average turn-around for iterating on changes of a feedstock. The same result occurred when merging to main, so it was not a one-off fluke either.

@velonica0 mentioned in https://github.com/conda-forge/conda-smithy/issues/2698 that
> We have OpenKylin machines with UltraRISC DP1000 boards that we'd like to connect through the RISE app, so the `ubuntu-24.04-riscv` label stays the same.

Do we have an idea where those machines land in terms of compute power w.r.t. to the current RISE runners? Also, how many OpenKylin machines are we talking about (vis-à-vis the 55 from RISE), and will the be a possibility to influence the runner choice, or will it just be one large pool? Having very divergent runner profiles in the same pool can be quite annoying, because CI jobs that passed fine a big runner might start failing if they randomly hit a slower one.

**Comment by luhenry** [comment]
> Do we have an idea where those machines land in terms of compute power w.r.t. to the current RISE runners?

I would expect at least a 3-4x speedup.

> Also, how many OpenKylin machines are we talking about (vis-à-vis the 55 from RISE)

I will let @velonica0 answer here.

> will the be a possibility to influence the runner choice, or will it just be one large pool?

We 100% have influence on runner choice. We can limit ourselves exclusively to the openKylin machines if there are enough. I wouldn't want to have 1 big pool where machines are chosen randomly since there is quite a big difference in terms of performance. We can have a tag to decide which one to have.

**Comment by velonica0** [comment]
Two DP-1000 units to start with, 8 cores and 16 GB RAM each, and we expect to add more later.

On the speedup, a measured number rather than an estimate: building coreutils-feedstock natively with the same recipe, the same image and `CPU_COUNT=4` to match the RISE runner's parallelism, the DP-1000 took 603 s against 1826 s on an EM-RV1, so 3x at identical parallelism. 

The remaining blockers on our side are network access and permissions rather than anything with the machines themselves. We're working through those and will hand the units over to RISE once they're sorted.

**Comment by h-vetinari** [comment]
Thanks a lot for the details @luhenry @velonica0! 🙏 

Even assuming a very optimistic quadrupling of the current 2 DP-1000 units (and successful integration through RISE), a situation of
* 55 EM-RV1 with ~9x slowdown compared to other arches
* 8 DP-1000 with ~3x slowdown compared to other arches

sounds to me like: we should have separate pools for these runners, and more importantly: that _**native compilation for riscv should be opt-in rather than the default**_, for feedstocks that require it, or would like to benefit from better testing. Otherwise we're just creating very tight bottlenecks all over conda-forge -- either by drastically longer runtimes or by congestion for the few available faster runners.

Just as a back-of-the-envelope guesstimate, I'd say that we'd need to have at least ~40-50 fast machines (DP-1000 or better) in a pool before going native by default becomes feasible.

CC @conda-forge/core

**Comment by velonica0** [comment]
As far as I know, the DP-1000 is currently the highest-performance hardware based on the RVA22 profile, and upcoming SoCs from major RISC-V hardware vendors will all adopt RVA23

Will the Conda community be open to supporting the latest RVA23 profile? After all, RVA23 includes the vector extension by default.

**Comment by h-vetinari** [comment]
If RVA23 makes a meaningful impact on the runtimes of builds, then I think it definitely should be discussed as an option. The overall calculus remains pretty much the same though: for native-by-default, we need 50+[^1] runners of _roughly_ comparable performance to what the other platforms take to run (and a slowdown by a factor of 3 is stretching "roughly comparable" to its limit IMO).

[^1]: or whatever number is necessary to not constantly have congestion across regular feedstock operations (assuming all widely-used feedstocks will end up supporting riscv), much less large-scale migrations.
